File: src/secrets.ts
// OS-native credential storage wrapper. Backed by `rezon_core::secrets`
// via the Tauri commands in `crates/rezon-web/src/secrets.rs`.
//
// Note there is no `keychainGet`. Keys are write-only from the
// frontend's side: the backend resolves them itself at request time
// (runtime override -> keychain -> environment), so the webview never
// holds a plaintext key and never puts one on an IPC payload. Use
// `keychainHas` when the UI needs to know whether one is configured.
//
// Account naming convention: `<purpose>:<scope>`. For cloud API keys,
// `api_key:<provider_key>` — e.g. `api_key:openai`. Must match
// `rezon_core::secrets::cloud_api_key_account`; renaming either side
// orphans whatever was saved under the old name.

import { invoke } from "@tauri-apps/api/core";

export async function keychainSet(
  account: string,
  value: string,
): Promise<void> {
  return invoke<void>("keychain_set", { account, value });
}

export async function keychainDelete(account: string): Promise<void> {
  return invoke<void>("keychain_delete", { account });
}

/// Whether a secret is stored, without retrieving it.
export async function keychainHas(account: string): Promise<boolean> {
  return invoke<boolean>("keychain_has", { account });
}

/// Convention helper: account name for a cloud-provider API key.
/// Use this rather than constructing `"api_key:"+key` inline so
/// renames stay centralized.
export function cloudApiKeyAccount(providerKey: string): string {
  return `api_key:${providerKey}`;
}

File: crates/rezon-web/src/secrets.rs
// Tauri command wrappers over `rezon_core::secrets`.
//
// The store itself lives in core so `rezon-tui` shares it and so key
// *resolution* (runtime -> keychain -> environment) is one policy in
// one place. This file is only the IPC surface.
//
// Note what is deliberately absent: there is no command that returns a
// stored key. The frontend can set one and ask whether one exists, but
// never reads the value back. A key the webview never holds is a key
// that cannot leak through devtools, a content-injection bug, or an
// IPC payload — and nothing in the UI needs the plaintext.

use rezon_core::secrets;

/// Write (or, with an empty value, delete) a secret.
#[tauri::command]
pub fn keychain_set(account: String, value: String) -> Result<(), String> {
    secrets::keyring_set(&account, &value)
}

/// Explicit delete. Idempotent on a missing entry.
#[tauri::command]
pub fn keychain_delete(account: String) -> Result<(), String> {
    secrets::keyring_delete(&account)
}

/// Whether a secret exists, without revealing it. This replaced a
/// `keychain_get` that handed the plaintext to the webview.
#[tauri::command]
pub fn keychain_has(account: String) -> Result<bool, String> {
    secrets::keyring_get(&account).map(|v| v.is_some())
}
